### ISSUE TITLE
helm reset --force reports OK but tiller is not deleted

### ISSUE THREAD

**Issue by marekaf** [helm/helm#4825]
We want to be able to do helm init --upgrade in a manual gitlab-ci stage 
helm init suceeded but didn't do anything, so I added helm reset --force followed with helm init, again, nothing happened 

when I do helm reset --force from my terminal and run the manual stage with helm init, everything works as expected, so the helm context in the pipeline is definitely set right.

 

```
$ helm version
Client: &version.Version{SemVer:"v2.11.0", GitCommit:"2e55dbe1fdb5fdb96b75ff144a339489417b146b", GitTreeState:"clean"}
Server: &version.Version{SemVer:"v2.11.0", GitCommit:"2e55dbe1fdb5fdb96b75ff144a339489417b146b", GitTreeState:"clean"}
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"10", GitVersion:"v1.10.7", GitCommit:"0c38c362511b20a098d7cd855f1314dad92c2780", GitTreeState:"clean", BuildDate:"2018-08-20T10:09:03Z", GoVersion:"go1.9.3", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"10+", GitVersion:"v1.10.7-gke.2", GitCommit:"8d9503f982872112eb283f78cefc6944af640427", GitTreeState:"clean", BuildDate:"2018-09-13T22:19:55Z", GoVersion:"go1.9.3b4", Compiler:"gc", Platform:"linux/amd64"}
$ echo "$HELM_HOME"

$ kubectl get pods -n kube-system | grep tiller
tiller-deploy-674ff75566-r5658                          1/1       Running   0          27m
$ helm reset --force
Tiller (the Helm server-side component) has been uninstalled from your Kubernetes Cluster.
$ kubectl get pods -n kube-system | grep tiller
tiller-deploy-674ff75566-r5658                          1/1       Running   0          27m
```
how is this still running when I get " Tiller has been uninstalled"?

Cloud Provider/Platform: 
GKE

**Comment by bacongobbler** [comment]
Okay. I'm not sure but it sounds like the pod was orphaned from the deployment. I have not been able to reproduce this locally but please let us know if you find out something.

**Comment by Tarkleigh** [comment]
I have the same problem. My use-case is to install some charts to a new cluster, then remove Tiller for security reasons with "helm reset --force". This worked like a charm with Helm 2.10 but does nothing with Helm 2.11. While it reports that Tiller has been removed successfully (even -v does not return an error), the pod is still there and will happily answer commands. A roll-back to 2.10 (and re-install of the Tiller so it is a 2.10 Tiller again) fixed the problem for now but of course this is not a permanent fix.

**Comment by bartimar** [comment]
@Tarkleigh if you need to install something and remove the tiller, maybe you can just use helm template and then kubectl apply without even having tiller in the first place?

**Comment by bartimar** [comment]
current workaround I'm using
```
$ kubectl get pods -n kube-system | grep tiller
tiller-deploy-674ff75566-r5658                          1/1       Running   0          2d
$ helm reset --force
Tiller (the Helm server-side component) has been uninstalled from your Kubernetes Cluster.
$ kubectl delete pods -n kube-system -l=name=tiller
pod "tiller-deploy-674ff75566-r5658" deleted
$ sleep 20
$ helm init --history-max=20 --debug --upgrade \ # collapsed multi-line command
```

**Comment by Tarkleigh** [comment]
@bartimar 
I tried helm template but had mixed success with it, it seemed highly dependant on the helm chart I was trying to install. The work-around should do for now but in any case this is a serious regression that should be fixed

**Comment by mattfarina** [comment]
You could use an approach like https://github.com/rimusz/helm-tiller or process documented on https://github.com/helm/helm/issues/4852#issuecomment-434368716

What version of Kubernetes are you using with Helm 2.11? Did you update k8s at the same time?

**Comment by Tarkleigh** [comment]
I used 1.11.3, I did not upgrade kubernetes at this time but it is possible that someone else did, I use a managed cluster. I can try again, if you want.

**Comment by abevoelker** [comment]
I can reproduce this consistently on GKE with both Helm 2.11 and 2.12

```
$ gcloud container clusters create "standard-cluster-1" \
  --zone "us-central1-a" --cluster-version "1.11.5-gke.4" \
  --machine-type "n1-standard-1"

$ gcloud container clusters get-credentials standard-cluster-1
Fetching cluster endpoint and auth data.
kubeconfig entry generated for standard-cluster-1.

$ kubectl create serviceaccount -n kube-system tiller
serviceaccount "tiller" created

$ kubectl create clusterrolebinding tiller-binding \
    --clusterrole=cluster-admin \
    --serviceaccount kube-system:tiller
clusterrolebinding.rbac.authorization.k8s.io "tiller-binding" created

$ helm init --service-account tiller
$HELM_HOME has been configured at /home/abe/.helm.

Tiller (the Helm server-side component) has been installed into your Kubernetes Cluster.

Please note: by default, Tiller is deployed with an insecure 'allow unauthenticated users' policy.
To prevent this, run `helm init` with the --tiller-tls-verify flag.
For more information on securing your installation see: https://docs.helm.sh/using_helm/#securing-your-helm-installation
Happy Helming!

$ helm repo update
Hang tight while we grab the latest from your chart repositories...
...Skip local chart repository
...Successfully got an update from the "stable" chart repository
Update Complete. ⎈ Happy Helming!⎈ 

$ helm version
Client: &version.Version{SemVer:"v2.11.0", GitCommit:"2e55dbe1fdb5fdb96b75ff144a339489417b146b", GitTreeState:"clean"}
Error: could not find a ready tiller pod

$ helm version
Client: &version.Version{SemVer:"v2.11.0", GitCommit:"2e55dbe1fdb5fdb96b75ff144a339489417b146b", GitTreeState:"clean"}
Server: &version.Version{SemVer:"v2.11.0", GitCommit:"2e55dbe1fdb5fdb96b75ff144a339489417b146b", GitTreeState:"clean"}

$ helm reset --force
Tiller (the Helm server-side component) has been uninstalled from your Kubernetes Cluster.

$ helm version
Client: &version.Version{SemVer:"v2.11.0", GitCommit:"2e55dbe1fdb5fdb96b75ff144a339489417b146b", GitTreeState:"clean"}
Server: &version.Version{SemVer:"v2.11.0", GitCommit:"2e55dbe1fdb5fdb96b75ff144a339489417b146b", GitTreeState:"clean"}

$ kubectl get pods -n kube-system | grep tiller
tiller-deploy-57f988f854-p4jft                                 1/1       Running   0          2m
```

Output of `helm version`:
```
Client: &version.Version{SemVer:"v2.11.0", GitCommit:"2e55dbe1fdb5fdb96b75ff144a339489417b146b", GitTreeState:"clean"}
Server: &version.Version{SemVer:"v2.11.0", GitCommit:"2e55dbe1fdb5fdb96b75ff144a339489417b146b", GitTreeState:"clean"}
```

Output of `kubectl version`:
```
Client Version: version.Info{Major:"1", Minor:"10", GitVersion:"v1.10.7", GitCommit:"0c38c362511b20a098d7cd855f1314dad92c2780", GitTreeState:"clean", BuildDate:"2018-08-20T10:09:03Z", GoVersion:"go1.9.3", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"11+", GitVersion:"v1.11.5-gke.4", GitCommit:"0c81dc1e8c26fa2c47e50072dc7f98923cb2109c", GitTreeState:"clean", BuildDate:"2018-12-07T00:22:06Z", GoVersion:"go1.10.3b4", Compiler:"gc", Platform:"linux/amd64"}
```

Cloud Provider/Platform (AKS, GKE, Minikube etc.): GKE

**Comment by whereisaaron** [comment]
I am seeing the same bug with helm 2.12.2 and GKE k8s 1.11.5-gke.5. Before `helm reset --force`:
```
$ k get all --namespace=helm
NAME                                READY     STATUS    RESTARTS   AGE
po/tiller-deploy-5f55c9bb7b-mppnb   1/1       Running   0          4m

NAME                TYPE        CLUSTER-IP     EXTERNAL-IP   PORT(S)     AGE
svc/tiller-deploy   ClusterIP   10.113.7.182   <none>        44134/TCP   4m

NAME                   AGE
deploy/tiller-deploy   4m

NAME                          AGE
rs/tiller-deploy-5f55c9bb7b   4m
```
After `helm reset --force` and _"Tiller (the Helm server-side component) has been uninstalled from your Kubernetes Cluster."_
```
$ k get all --namespace=helm
NAME                                READY     STATUS    RESTARTS   AGE
po/tiller-deploy-5f55c9bb7b-mppnb   1/1       Running   0          4m

NAME                          AGE
rs/tiller-deploy-5f55c9bb7b   4m
```
Only the Service and Deployment are deleted. When deleting the Deployment, cascade delete appears to have been disabled when helm deleted the Deployment.

If I delete the deployment myself with `kubectl` is works just fine
```
$ k --namespace=helm delete service tiller-deploy
service "tiller-deploy" deleted
$ k --namespace=helm delete deployment tiller-deploy
deployment "tiller-deploy" deleted
$ k get all --namespace=helm
No resources found.
```

So it appear `helm` is doing something that actually disable cascade delete, the equivalent of:
```
$ k --namespace=helm delete deployment tiller-deploy --cascade=false
```
'helm reset` works just like `--cascade=false` is set:
```
$ k get all --namespace=helm
NAME                                READY     STATUS    RESTARTS   AGE
po/tiller-deploy-5f55c9bb7b-pw9lb   1/1       Running   0          45s

NAME                TYPE        CLUSTER-IP     EXTERNAL-IP   PORT(S)     AGE
svc/tiller-deploy   ClusterIP   10.113.12.46   <none>        44134/TCP   45s

NAME                   AGE
deploy/tiller-deploy   45s

NAME                          AGE
rs/tiller-deploy-5f55c9bb7b   45s
$ k --namespace=helm delete deployment tiller-deploy --cascade=false
deployment "tiller-deploy" deleted
$ k get all --namespace=helm
NAME                                READY     STATUS    RESTARTS   AGE
po/tiller-deploy-5f55c9bb7b-pw9lb   1/1       Running   0          1m

NAME                TYPE        CLUSTER-IP     EXTERNAL-IP   PORT(S)     AGE
svc/tiller-deploy   ClusterIP   10.113.12.46   <none>        44134/TCP   1m

NAME                          AGE
rs/tiller-deploy-5f55c9bb7b   1m
```

**Comment by pytimer** [comment]
I also have the same bug with helm 2.12.1 and kubernetes version is 1.11.6, like @whereisaaron said.

**Comment by knabben** [comment]
I'm interested on this one.

**Comment by Tarkleigh** [comment]
Is there any update on this? I just tested it with release v2.12.3 and the bug is still there.

**Comment by bacongobbler** [comment]
Good call @Tarkleigh. Promoted to a bug.

**Comment by AWDDude** [comment]
I have this problem as well. Is there a workaround to get rid of the stuck pod? If I delete the pod it just gets recreated.

**Comment by bacongobbler** [comment]
Try deleting the replicaset or the deployment managing the pod.

**Comment by AWDDude** [comment]
Yeah deleteing the replica set did it. I coulda sworn I checked if there was a replica set earlier I must not have been looking in the right namespace

**Comment by bacongobbler** [comment]
I just confirmed this bug as well with 2.12.3.

```
PS C:\Users\me\code\go\src\k8s.io\helm> helm version
Client: &version.Version{SemVer:"v2.12.3", GitCommit:"eecf22f77df5f65c823aacd2dbd30ae6c65f186e", GitTreeState:"clean"}
Server: &version.Version{SemVer:"v2.12.3", GitCommit:"eecf22f77df5f65c823aacd2dbd30ae6c65f186e", GitTreeState:"clean"}
PS C:\Users\me\code\go\src\k8s.io\helm> helm reset --force
Tiller (the Helm server-side component) has been uninstalled from your Kubernetes Cluster.
PS C:\Users\me\code\go\src\k8s.io\helm> helm version
Client: &version.Version{SemVer:"v2.12.3", GitCommit:"eecf22f77df5f65c823aacd2dbd30ae6c65f186e", GitTreeState:"clean"}
Server: &version.Version{SemVer:"v2.12.3", GitCommit:"eecf22f77df5f65c823aacd2dbd30ae6c65f186e", GitTreeState:"clean"}
PS C:\Users\me\code\go\src\k8s.io\helm> kubectl -n kube-system get pods
NAME                                   READY     STATUS    RESTARTS   AGE
heapster-779db6bd48-628n9              2/2       Running   0          51m
kube-dns-autoscaler-5b94d96445-brpr2   1/1       Running   0          51m
kube-dns-v20-6df5d5c657-7blzp          4/4       Running   0          51m
kube-dns-v20-6df5d5c657-97c29          4/4       Running   0          47m
kube-proxy-hghlk                       1/1       Running   0          48m
kube-proxy-j6ktl                       1/1       Running   0          48m
kube-proxy-n46f5                       1/1       Running   0          48m
kube-svc-redirect-9gznt                2/2       Running   0          48m
kube-svc-redirect-dvfxj                2/2       Running   0          48m
kube-svc-redirect-f78jb                2/2       Running   0          48m
kubernetes-dashboard-f49587769-8xfzv   1/1       Running   0          51m
tiller-deploy-66cdfd5bc9-mf4bf         1/1       Running   0          47s
tunnelfront-6b86fc559b-c4j7h           1/1       Running   0          51m
```

I'll have a look at https://github.com/helm/helm/pull/5161 and give that a go.

**Comment by bacongobbler** [comment]
fixed via #5161. Closing!
